### ISSUE TITLE
Discussion around command line parser changes

### ISSUE THREAD

**Issue by cmnbroad** [broadinstitute/picard#786]
We propose to make changes to Picard to also support Posix-style command line syntax in addition to the existing syntax. The new syntax would be enabled by default, but the user would have the ability to opt-out and use current syntax by using an alternate jar file, which will be provided pre-compiled with each release. In the event the user opts out, the existing command line syntax will remain unchanged and be fully backward compatible.

Following is a Posix-style command line:

> java -jar picard.jar ValidateSamFile -I input.bam -MODE SUMMARY

The equivalent Picard command line using the current syntax is:

> java -jar picard.jar ValidateSamFile I=input.bam MODE=SUMMARY

**Overview of the proposed changes:**

- Change Picard to use the command line parsers and annotations in the (BSD 3 licensed) [Barclay](https://github.com/broadinstitute/barclay) library. Barclay contains two command line parsers (a slightly modified version of the existing Picard parser, and a posix-style parser based on jopt-simple). Both parsers use the same set of argument annotation classes, also defined in Barclay.
- Change picard.jar to use the new command line parser, and add a new build target (alternate jar) that would use the existing parser.

**Picard source changes:**

- Add a dependency on Barclay.
- Change all occurrences of the @Option, @CommandLineProgramProperties, and @PositionalArguments annotations to use equivalents in the Barclay package “org.broadinstitute.barclay.argparser” package (named @Argument, @CommandLineProgramProperties, and @PositionalArguments respectively). This is a fairly easy package name change that can be done automatically with a typical IDE.
- Add a command line translator that converts old-style command lines into posix-style for purposes of test execution. Tests would continue to be written using the old style syntax for now (recognizing both could be done down the road), but the test would be run once as they currently are and once with the translator and new parser.

**Note on `overridable` Arguments**

The Picard parser currently allows a tool to duplicate and override arguments (via the @Option attribute `overridable`) that are inherited from a base class in order to change either the default value or the values of other Option attributes, such as the doc string or the “required” attribute. This is used by several tools for example to make the reference argument, which is usually optional, required for that tool. We propose to remove this feature of the argument parser, and to replace it with equivalent functionality in the form of @ArgumentCollection annotations. The @ArgumentCollection feature allows a tool to have a reference (annotated with @ArgumentCollection) to an object that itself contains one or more embedded @Arguments. This allows the tool to select the attributes of its arguments by dynamically selecting the class to instantiate, (i.e., one of RequiredReference or OptionalReference).

**Summary**

**End Users:**

- Users of Picard who choose to not use the new syntax can simply use the alternate jar that implements the syntax currently in use.

**Developers:**

- Downstream projects that depend on Picard will need to make a one-time source code change to reference the new Barclay package, annotation, and annotation attribute names.
- Existing tools will need to be changed to use @ArgumentCollection place of override arguments: (For Picard itself, this includes CollectHsMetrics, CollectAlignmentSummaryMetrics, CollectRbsMetrics, CollectWgsMetrics, CollectWgsMetricsFromSampledSites, CreateSequenceDictionary MergeBamAlignment, LiftOverVCF).

**Documentation:**

- The tool documentation will be updated to reflect the new POSIX syntax only.
- Usage examples will be rewritten to show the new POSIX syntax.
- A migration guide will be published detailing how to make the changes outlined above for any projects that need to do so. The changes relating to override arguments will be highlighted.

**Comment by nh13** [comment]
A few thoughts on the top of my head other to say yay:
1. would `REFERENCE_SEQUENCE` turn into `--reference-sequence`?  It looks like underscores are not allowed according to [this source](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html).  Also, it would be nice not to have all upper-case arguments.
2. what's the reasoning behind the opt-out use of the new style, since A LOT of people have code surrounding the old syntax.  A reasonable compromise would be for the default jar (`picard.jar`) to use the old syntax for 3-6 months (some fixed period of time) for folks to make the transition while having _both_ in the wild.
3. I am not sure I understand the `@ArgumentCollection` so could you post a diff of one of the tools showing it using `@Option` and then `@ArgumentCollection`?
4. Given this is a **major** change to arg parsing, are there other arg-parsing features that we would like to request (ex. `@<file>` would automatically read in the given file full of command line arguments, or `-i A B C` is the same as `-i A -i B -i C`)?

**Comment by magicDGS** [comment]
Some answers to your questions, @nh13, after my experience using Barclay:

1. I have arguments in the format `--reference_sequence`, so underscore are allowed in the correct behavior. Regarding your question about upper-case arguments, both using the field name or the `fullName` from `@Argument` could include lower and upper-case values.
2. I do not have any strong opinion about this.
3. I will add an example of this later.
4. In the new parser, there is an special argument (`--arguments_file`), which reads a file full of arguments and apply them to the program, and other arguments could be specify outside. In addition, there is a new new behaviour for the `Collection` arguments, which allows to provide a file with `.list` extension for read in the elements. I think that there is no support for `-i A B C` instead of `-i A -i B -i C`, but @cmnbroad will know better.

I also have several questions/comments about this:

1. One difference between the "legacy" cmd parser and the new one is that the `printOrder` for the later is ignored. I would like to have a better control on how the arguments are displayed in the help, but more details are bellow.
2. As I pointed out in Barclay, there is no annotation or way to cluster together several arguments and add a header to explain what are they related for. This remove the support for `@NestedOptions` annotations. It is true that it's unused in Picard, but I have some code using them and, at the same time, I would like to keep this behaviour. Is there any plan to implement this in Barclay soon?
3. Collections are handled in a different way currently, because they replace the defaults instead of append. Anyway, this is going to have a switch if I remember correctly. Is this switch specified for every collection (as an `@Argument` field) or for the parser? I would prefer the first one, because it will allow to have control over it depending on the situation.
4. I think that @cmnbroad forgot to mention that another change in Picard is `CommandLineProgramGroup`, which is also defined in Barclay now.

All this said, for my point of view this is an improvement for Picard because there are also new features that may be useful, like the `@Hidden` or  `@Advance` annotation to hidde or treat specially some arguments, the possibility of have tagged arguments, and `@ArgumentCollection` containers for changing documentation/visibility of arguments.

**Comment by magicDGS** [comment]
Example of argument collection for the `REFERENCE_SEQUENCE` argument:

```java
public abstract class CommandLineProgram {
    ...
    // @Option(shortName = StandardOptionDefinitions.REFERENCE_SHORT_NAME, doc = "Reference sequence file.", common = true, optional = true, overridable = true)
    // public File REFERENCE_SEQUENCE = Defaults.REFERENCE_FASTA
    @ArgumentCollection
    public ReferenceArgumentCollection referenceArg = getToolReferenceArgument();

    public ReferenceArgumentCollection getToolReferenceArgument() {
        return new OptionalReferenceArgumentCollection();
    }
    ...
}
...
public interface ReferenceArgumentCollection {
    public File getReferenceFile();
}
...
public final class OptionalReferenceArgumentCollection extends ReferenceArgumentCollection {
    @Argument(shortName = StandardOptionDefinitions.REFERENCE_SHORT_NAME, doc = "Reference sequence file.", common = true, optional = true)
    public File REFERENCE_SEQUENCE = Defaults.REFERENCE_FASTA

    public File getReferenceFile() { return REFERENCE_SEQUENCE; }
}
...
public final class RequiredlReferenceArgumentCollection extends ReferenceArgumentCollection {
    @Argument(shortName = StandardOptionDefinitions.REFERENCE_SHORT_NAME, doc = "Reference sequence file.", common = true, optional = false)
    public File REFERENCE_SEQUENCE;

    public File getReferenceFile() { return REFERENCE_SEQUENCE; }
}
```

Classes requiring the reference parameter should override the `getToolReferenceArgument()` method. Even if there is no need for this argument, they could create a class that does not have the parameter and remove it.

Finally, tools should use the `referenceArg.getReferenceFile()` to use the file and/or check if it is null.

**Comment by nh13** [comment]
@magicDGS Thanks for the answers and this is definitely an improvement, but I certainly think that we should agree on the parsing behaviors, for example I would prefer long options not to have underscores and my reading of GNU's long options agrees with my preference [source](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html).  Other examples include replacing a list with defaults with an empty list, and using `Optional` rather than `null`.  I would be tempted to look for some useful ideas from python's [argparse](https://docs.python.org/3/library/argparse.html) library, which I find very full featured.

I also think that my point (2) is very important to allow pipeline authors to update how they call Picard as well as others to update how they build or extend Picard.

**Comment by tfenne** [comment]
@cmnbroad @magicDGS: I'm in two minds about this.  It does seem like an improvement overall, but I also have a _lot_ of code that depends heavily on Picard that I feel is going to cause me an awful lot of pain when the day comes that I need to upgrade to the new version that implements this change.  That code falls into two broad categories:

1. Pipelines that use Picard tools, invoking them via the command line
2. Toolkits of custom tools that I've built that use Picard's command-line processing framework to build a picard-like JAR

I'm particularly concerned about under the cover changes that will break compatibility due to subtly different behaviours between Picard's current arg parsing and Barclay's.  E.g. if it is the case that the way collection arguments with default values are treated when values are specified on the command line too is changing, that's going to be nightmarish to track down and fix usages.  

I haven't spent much time looking at Barclay yet, but from a couple of conversations I've had and examples I've seen, it sounds like it's more or less a cleaned up and extracted version of the GATK's parsing framework that has been extended to support picard-style too?  I'll echo @nh13 that if Picard is to make a switch to GNU style, it would make it an easier sale (to me at least) if it:

1. Really strove for GNU style.  I.e. long arguments with double dashes and words separated by dashes (not underscores, not camel case), and flags with single letters and a single dash.  Support for `--in 1.bam 2.bam 3.bam` style usage, so that in a shell one can write `--in *.bam`.  Support for `-imy.bam`, `--in my.bam` and `--in=my.bam` formats.
1. Adopted the use of, or at least provided strong support for, `Optional` as @nh13 suggested, so that optional arguments can be specified using default values of `Optional.empty()` instead  of `null`, and without having to set a flag on the annotation to state a value is optional
1. Supported some of the better ideas from Python's `argparse`.  Specifically they way argparse does argument files is vastly superior IMHO, and we've adopted that in our work.  In argparse any argument that starts with a specific special character (`@` by default) is expected to be a file path to a file with a list of arguments in it.  You can have many on the command line, and they are substituted at the point they appear.  So you can write things like `--ref foo.fa @normalbams.txt @tumorbams.txt`.
1. A really comprehensive migration guide is written that covers _all_ the things that might surprise someone during migration, both as a user of command line picard tools and of Picard's old APIs for building toolkits.

**Comment by tfenne** [comment]
One other thought on the "feature wish list".  I can't remember if we've said it here before but @nh13 and I implemented our own [arg parser](https://github.com/fulcrumgenomics/dagr/tree/master/sopt) in scala a while back, and use it heavily in [fgbio](https://github.com/fulcrumgenomics/fgbio).  One of the features we've been kicking around that I think we will end up working on soon, and could perhaps collaborate with you guys on to have in Barclay also is support of Markdown for all documentation strings (usage and argument).  The idea would be to use Markdown everywhere and _not_ HTML, and then have support to go from Markdown -> HTML for website use and Markdown -> Plain Text for output to a terminal, possibly using ANSI escape sequences to render some features like bold, italic etc.

Given how much HTML is in Picard usage statements now and the hacky way it is stripped out, this would be a nice improvement.  If you're interested, we could probably work together on patching a Markdown library to add plain-text and ANSI-escaped translations.

**Comment by magicDGS** [comment]
I like your idea for documentation, @tfenne! I would collaborate there if you do something for that!

**Comment by cmnbroad** [comment]
Thanks for all the lively feedback. Some more thoughts and answers to some questions:

- Barclay contains two separate parsers. They are completely independent from each other, other than relying on the same set of shared annotation classes. One of those parsers *is* the Picard parser, modified slightly to use the Barclay annotations. There should be no runtime transition pain for any end user or pipeline consumer that continues to use that syntax.
- I think the idea of having a transition period of a couple of months before we switch the defaults is reasonable.
- Although there currently is a difference between how the two parsers treat collection arguments (append vs replace), as part of this change we plan to make them have behavior identical to the current Picard behavior when used in Picard.
- @magicDGS posted a good example of using an ArgumentCollection - let me know there are more more questions about that.
- Underscores are allowed; both parsers work with the existing argument names as they are in Picard (there is no case translation or other differences that I know of). The new parser accepts “POSIX style” command lines, but it wasn’t a goal to be aggressively POSIX "compliant".
- A lot of these ideas mentioned in this thread would make great enhancements going forward (like @Optional, or using markdown, both of which I like), but I think those can and should be part of the evolution of Barclay going forward. We’d like to start with what we have since we’ve done a lot of work to get the current point where we have full compatibility, and move forward.

**Comment by yfarjoun** [comment]
Thanks for this @cmnbroad! 

I recognize support the effort to make this effort towards both "new and better" and "not breaking what works", both of which are important aspects!

👍

**Comment by yfarjoun** [comment]
Thinking about this more over the weekend, I think that there are some suggested changes around POSIX compliance that would be a breaking change if implemented later (and thus would probably _not_ be accepted as a later PR)...in particular, I'm thinking of the single character flag option, and the casing/underscore vs. dash issues. 

1. Casing + dash vs. underscore: Thinking forward to when picard tools will be exposed via GATK it would be good to similar parameters, not `REFERENCE_SEQUENCE` in one tool and `reference` in another. This can be enabled by lowercasing the parameter names in the POSIX implementation, converting `_` to '-' *and* allowing unambiguous substrings of parameters. Thus --reference would mean the same in both POSIX picard and GATK4. 

2. Perhaps the @Annotation could have a `flag` parameter (distinct from `shortName`) that would accept only single-character arguments, and would allow any of space,=,emptyset separator. This would imply that both longname and short-name would be used with a double dash and only the most commonly used flags will be "allowed" to have flags as they are more prone to clashing. 

While these would clearly imply more work on Barclay, I think that if we want to do them, we should perpare for them before the beta release so that the breaking changes that they imply do not create a impediment. Given that we are already accepting an eventual breaking change in picard and that GATK4 is still in alpha, this might be the only time to do these sorts of changes.

**Comment by cmnbroad** [comment]
A couple things:

I did a few tests using the GATK parser to get a level set on whats accepted now. For an @Argument defined with the following attributes (note that "fullName is an attribute on the Barclay @Argument annotation; and unlike Picard, this overrides field name):

```
fullName="reference"
shortName="R"
```

the following command line (I'm including only relevant fragments) are all accepted by the current Barclay implementation:

```
-R ref.fasta
-Rref.fasta
-R=ref.fasta
--R ref.fasta
--R=ref.fasta
--reference ref.fasta
--reference=ref.fasta
```

The Barclay GATK parser itself is built on top of https://pholser.github.io/jopt-simple/, but is a bit of a hybrid in that it currently implements some features (like required-ness) itself instead of using the jopt-simple support for it (I'm not sure why - this was before my time). Jopt-simple also implements short arg clustering. However, I think the implementers made a conscious decision not to support multiple args for an option (`-i A B C`), which is unfortunate since as @tfenne and @nh13 mentioned, it precludes using shell expansion.

On the subject of getting consistency between Picard and GATK, one alternative option would be to employ the "fullName" attribute in the Picard source for the purpose of getting long name parity with GATK (as an example, add `fullName="reference"` to the REFERENCE_SEQUENCE annotation in Picard). That attribute is ignored by the Picard parser, but is used by the GATK parser as I mentioned above. So using the Picard parser, you'd have "R" and "REFERENCE_SEQUENCE", and using the new parser you'd have "R"  and "reference" (which matches GATK). It doesn't solve all of the cases (I'm not sure conversion does either), but seems like a less error prone way to cover the common cases where the short names are already aligned (like input, output, reference).

**Comment by nh13** [comment]
@cmnbroad I really like shell-expansion that we have in [fgbio](https://github.com/fulcrumgenomics/fgbio) and [dagr](https://github.com/fulcrumgenomics/dagr).  I am assuming the time has passed to brainstorm which parser to use, which is a bit of a shame, but understandable given the work you have already done.

**Comment by lbergelson** [comment]
1. **Markdown instead of html**: HTML is a terrible thing to have inline in code.  I would love to change this.  I don't know how much work it would be.
2.  A lot of the features/lack of features in the existing arg parser are based on a desire to deviate as little as possible from the existing picard/gatk parsers.  There's no reason we couldn't revise some of those. ex:
-  **Unambiguous abbreviations**.  We actually had this in gatk, but it was causing confusion because the differences in argument lists between tools made the meaning of arguments change in unexpected ways.  I.e `-V` was fine for verbosity in most tools, but suddenly it meant `variant` in others. We actually patched jopt to add support for disabling this feature.
-   **`-I a b c` vs `-I a -I b -I c`**:  I believe jopt supports this feature but we explicitly didn't enable it in gatk because it's not how it traditionally worked.  I'd be in support of enabling it.  It does cause parsing ambiguities if you allow it in concert with positional arguments, so we'd have to think about

3.   **Arguments Files**:  Barclay already supports argument files with the special argument `--arguments_file`  It works similarly to how you describe it.  It identifies all `--arguments_file` arguments, and adds those arguments to the list of args for parsing.  It doesn't maintain their relative position to the other arguments.  We could probably change it.  Having an explicit argument rather than a mysterious character `@` seems better to me.  (although `--arguments_file` is a mess of a name, we could do better.  Do you use the positionality of it in a meaningful way?  I.e. do you ever do `--input @listOfInputs`?  We would use an `.list` file for that instead.  

4.  **Flags**: We already support flags in gatk.  Any boolean option can be specified with or without an argument.  If specified in this way it's treated as true.  You can use single character short names and specify them adjacent to each other if you want. 

5.  **Optional**:  Supporting this would be great.  

6. Naming disparities.  I'm not against disallowing `_` in long names, but it would be a change for gatk and we'd have to rename a lot of parameters I think.  Automatically translating `_ -> -` seems like a recipe for confusion though.  One thing we could do is just add an additional name field  `shortName` `fullName` `alternativeName`.  Jopt supports arbitrary numbers of argument names so it would be an easy and relatively painless addition.

**Comment by vdauwera** [comment]
(following Louis' numbered list to preserve everyone's sanity)

1. **Markdown instead of html** - I would looooove this so much. I hate to write html as much as you hate seeing it in the code. For the web-based docs this should be trivial; we already use markdown parsing abundantly on the GATK website at least. For the CLI help I don't know but I'll eat my hat if there aren't appropriate libraries just begging to be used for this. 

2. **Features/lack thereof** - Indeed, we aimed to stick as closely as possible to the current state so as to minimize migration pain.

    - **But let's keep abbreviations as unambiguous as possible**
    - I have no strong feelings on **-I a b c vs -I a -I b -I c** except that I plan to use a single set of conventions in our docs for the sake of sanity, and generally speaking, having more different ways of doing the same thing ends up being more confusing than helpful, especially to beginners, who are the people I worry about the most.

3. **Arguments Files** - I'm also skeptical of the mysterious `@` character and prefer obvious, explicit things like `--arguments_file`, albeit perhaps with a better name (not that I'm terribly bothered by this one).

4. **Flags** - Nothing to add

5. **Optional** - Nothing to add

6. **Naming disparities** - Oy, renaming lots of parameters would cause a **lot** of migration pain, which we were specifically trying to avoid. I favor encouraging (or possibly enforcing) standard conventions in new development going forward, but trying to rename lots of things (including through automatic translation) sounds like a recipe for disaster, and out of proportion considering the problem is essentially a cosmetic one (I would love parameters to all be nice and uniformly formatted, and I would also like a pet unicorn). And I'm not crazy about adding yet another option for names because from the support point of view that's an additional permutation to store in memory when we're parsing user command lines and trying to figure out what they're doing. Not sure I want to take on that extra complexity just in the name of aesthetics.

**Comment by yfarjoun** [comment]
Regarding 6. ( I think that this is the only point that we disagree AND it would be problematic if we do not resolve it now)
 
In GATK4, there is no migration cost, we can change the arguments as we like since we are pre-beta. Let's get them right the first time around and decide on a sane choice of naming convention (I do like the POSIX lower-case-separated-by-dashes-for-long-names and single characters for a select few single character use-cases, only for the most commonly used ones like input, output, reference, variant, etc) 

This will enable us to have a matching naming scheme in picard (using another field in the annotation) which we could eventually drop if the transition to POSIX completes with no casualties.

**Comment by lbergelson** [comment]
I should point out that a lot of the features I mentioned are only supported in the new barclay parser, not the implementation of the legacy picard parser.  It would be a lot of work to add them to the picard parser.  (The new style parser has some additional useful features that weren't mentioned here as well... )  I't

@yfarjoun there's actually quite a large migration cost for changing gatk4 options.  They've been kept as close to gatk3 as was reasonable, so changing them is akin to changing the picard option names.

**Comment by yfarjoun** [comment]
well, I figured that if we are accepting a syntax change for folks using
Picard, we could share the love/pain and create a more consistent naming
convention in GATK4. Given that it's a major version change and the
invocation is different, I don't think that the naming of parameters is
going to be the issue....but hey, what do I know?

**Comment by tfenne** [comment]
@lbergelson I think that's understood.  The long term goal is, as I understand it, to move both GATK and Picard to a single option/argument flavor, which would be the GNU/posix style in Barclay, with the Picard style implemented to provide a smoother migration path.

I'll try and answer a few questions (using bullets so as not to confuse numbers ;) ):

* Regarding MarkDown, there are plenty of good libraries out there to parse markdown and to turn it into HTML, but surprisingly not really any to turn it into reasonably well laid out plain text.  I've spent some time prototyping this for `sopt`, the scala option parser that @nh13 and I use.  It's in scala, but the real hard work was understanding the AST that the flexmark library generates - I think the ~200 lines of scala could be converted to ~400 lines of scala ( 😛 ) pretty easily.  Take a look at https://github.com/fulcrumgenomics/dagr/pull/279 .  It parses GitHub flavor MarkDown (without some of the extensions like tables) and produces reasonably nice looking plain text wrapped to a user-chosen line width.
* The `--in a b c` syntax is just super useful in general.  I agree that the `.list` file support can help since you can do `ls -1 *.bam > bams.list; gatk --in bams.list`, but it's nice to be able to do that in a single step.  It makes interactive use so much easier.
* The reason we followed argparse's use of a special character for argument files is mostly just to avoid introducing special options that then have to be documented for every program.  So long as the mechanism supports multiple files I think the two are equivalent.  You can imagine having argument files for resource bundles (b37 and b38) and caller options, and running something like `gatk HaplotypeCaller --argument-file b38.args --argument-file aggressive-downsampling.args --in ...`
* On the strict naming (underscores vs. hyphens etc.) it would definitely be good to try and enforce this going forward.  I think it would be worthwhile for the Picard tools auto-translating underscores to hyphens, since the argument naming in Picard is all `WORDS_SEPARATED_BY_UNDERSCORES`, it would be nice if Picard could be autotranslated to `--words-separated-by-hypens`.

**Comment by tfenne** [comment]
Two more thoughts:

* On enforcing short names being single characters.  I personally think this is a good idea as it makes it possible to later add support for things like `-abc` which turns on options `-a` `-b` and `-c`.  But it's also partly a reaction to things like this in the GATK:
    - `-fixMisencodedQuals,--fix_misencoded_quality_scores`
    - `-allowPotentiallyMisencodedQuals,--allow_potentially_misencoded_quality_scores`
    - `-globalQScorePrior,--globalQScorePrior`
    - `-writeFullFormat,--never_trim_vcf_format_field`
where we get either two long names (but one with `-` and one with `--`), or even the same long name written out twice in the usage
* If the switch to Barclay made it so that the GATK printed the default values for it's arguments....  

![image](https://cloud.githubusercontent.com/assets/1609210/24934001/42b6f824-1ee7-11e7-8f99-d4373f3bbd89.png)

**Comment by droazen** [comment]
@tfenne GATK4/barclay does indeed print out the default values for its arguments in the `--help` output! :)

One question that hasn't come up yet, but I think would be good to attempt to answer as part of this discussion is how long the legacy/non-POSIX arg parser needs to stick around in barclay. Months? A year? Forever? 

The reason I ask is because it seems like people have a lot of wishlist-type features that they want to see added to barclay over time, and having to keep the legacy parser around indefinitely might limit/impair our ability to make certain improvements. It was actually surprisingly difficult to get to a state where the two parsers can be toggled between transparently on the same source, without breaking strict compatibility with the old syntax/behavior when using the old parser and without users having to worry about subtle/problematic differences in behavior (as opposed to useful new features!) when using the new parser -- maintaining this state of affairs indefinitely seems hard. At the same time, we definitely want to keep the old parser around long enough to minimize migration pain for the community.

Thoughts on this issue?

**Comment by magicDGS** [comment]
I think that using the old parser in Picard should not be maintain too long, neither in Barclay. Some features that I've implemented already have not support in the legacy style, and it just print a warning that it is not useful for the final user. For instance, hidden and/or advance annotations are not handled, neither boundaries for numeric parameters.

Maybe a good idea is to maintain in Picard the old-style parsing for a couple of releases before removing it, and announcing that it will disappear in the near future. I did something similar with one of the tools used in my team when I changed the parsing to Barclay....

**Comment by tfenne** [comment]
@droazen That's great news.  That's been probably my biggest usability complaint about the GATK for a long time, and it will be great to see that go away!

I would suggest that the legacy Picard format stick around for somewhere between 6-12 months.  I can't recall, but I feel like that's about inline with the deprecation policy.  I think it's important to give people sufficient time that they can plan their own migration when it works best for them, and not be rushed into it if, e.g. a critical bug fix comes out shortly after the switch to Barclay, but that "forever" would incurr too much of a support burden.  Maybe it could even be something like:
* The next three major releases (~3 months) will support both
* Any critical bug fixes (fingers crossed, maybe there won't be any?) would be back-ported to, and point releases made from. the last release that supported old-style arguments for another ~6 months

**Comment by droazen** [comment]
@tfenne A 6-12 month period for maintenance of the legacy parser sounds reasonable to me -- do the others in this thread agree? Perhaps that could translate into "end of 2017"?

Think I would favor a straightforward 6-12 month transition period over 3 months followed by 6 months of backporting critical fixes, as in my experience backporting bug fixes can itself be a source of bugs :)

**Comment by cmnbroad** [comment]
It sounds like there is a lot of support for a number of features discussed here, so I'm going to make Barclay tickets for those. Let me know if I missed any:

- Markdown (we've talked about this for a while - maybe we can benefit from some of the [dagr](https://github.com/fulcrumgenomics/dagr/pull/279) work)
- Unambiguous abbreviations
- Support for option arguments
- Improve argument file support
- Optional (we already have a ticket for this somewhere)

Short term I think that leaves two related issues: case/underscore/hyphen translation and Picard/GATK similarity. We have quite a bit of work to do in a short amount of time just to make this integration happen at all, even without addressing these, and I'm struggling to find a way to mutually reconcile all of our collective goals (we all want to maintain compatibility for Picard standalone, at least medium term; we want Picard in GATK4 to be similar to GATK4, GATK4 to be similar to GATK3, and ideally have GATK4/Picard to be POSIX-y).

Nevertheless, we *do* need to resolve these remaining issues. I propose we should continue to discuss, but I'd like to not have forward progress be contingent on it (also, it will be a couple of weeks before I'm ready with a PR anyway).

**Comment by droazen** [comment]
@cmnbroad:

* I'd add a ticket to see whether we can support `--in a b c` syntax in barclay as well, since that would be useful and a lot of people have suggested it.

* I would remove the abbreviations ticket from that list, since as @lbergelson mentioned above we had this feature on for a while in GATK4 and it was a source of major headaches:

```
We actually had this in gatk, but it was causing confusion because the differences in argument lists between tools made the meaning of arguments change in unexpected ways. I.e -V was fine for verbosity in most tools, but suddenly it meant variant in others. We actually patched jopt to add support for disabling this feature
```

Could you weigh in on the proposal above of maintaining the legacy parser for a 6-12 month period in barclay before retiring it, @cmnbroad (since you would likely be one of the people maintaining it)? I'd like to make sure everyone is ok with that proposal before the discussion returns to the underscore vs. hyphen debate.

**Comment by cmnbroad** [comment]
Yes, my line item "Support for option arguments" was supposed to say "Support for multiple option arguments", by which I mean support for `--in a b c`. I edited that entry to be clearer, and will not make a ticket for ambiguous args.

I think the 6-12 month timeframe sounds about right to me. Hopefully we can keep the maintenance to a minimum.

**Comment by droazen** [comment]
I asked @lbergelson , @yfarjoun , and (indirectly) @ktibbett in private about the proposal of a 6-12 month lifetime for the legacy arg parser in barclay, and they all gave thumbs up to the idea provided that the 6-12 month period starts upon release of the first Picard version that depends on barclay, and not sooner, which makes complete sense.

So I guess we can return to the final sticking point of underscores vs. dashes and how much we want to try to clean up our argument names as part of this proposal. I met with @cmnbroad and @vdauwera to try to figure out how much appetite there is for this sort of thing on the GATK side. We settled on the following proposal:

* Before the release of GATK4, we will systematically standardize all of our "long" argument names to obey the `lowercase-with-dashes-as-separators` convention, with an opt-in check in barclay (on for GATK, off for Picard) that will raise an error if this convention is violated.
* We will fix the more absurd cases of inconsistent short vs. long names, like @tfenne 's embarrassing `-writeFullFormat,--never_trim_vcf_format_field` example.
* We will not require short names to be single letters, however. The use case of "-abc which turns on options -a -b and -c" that has been mentioned above, while certainly very standard, is not really useful for GATK, since the single-character arg names are typically reserved for engine-wide input arguments (eg., `-R`, `-I`, `-V`, `-L`, etc.) rather than flags/options. We also really want to preserve convenient two-character short names like `-RF` for `--readFilter` that our users are used to.
* Although we will not require short names to be single letters, we will try to be more restrained/disciplined  in our use of short names, and eliminate cryptic short names for little-used/obscure arguments.

So this is not quite a "pure" solution, but it does address most of the complaints about the past sloppiness of GATK argument names, I think, while also taking into account the needs of GATK users.

As for Picard, obviously straight-up argument renaming like we're doing in GATK4 is off the table. Of the proposals above, I think I like the "add a fullName to all Picard args, and accept fullName, shortName, or the traditional 'name equal to the field name' for all args" option best, as the dynamic translation/renaming option seems a bit sketchy to me. But I defer completely to the Picard community and @vdauwera on this issue.

**Comment by tfenne** [comment]
@droazen et al. thank you for taking the feedback and running with it.  I think the proposal you've come up with is a really good compromise, and I'm happy to support it.

Putting my Picard hat back on... since nobody is using Picard with GNU/Posix style arguments today I do think there's a fair amount of wiggle room to get things right the first time on that side.  I'm curious what @ktibbett, @yfarjoun, @nh13 and others think, but my take is that a simple auto-translation of field names like `field.getName().toLowerCase().replace('_', '-')` would actually end up with very sensible looking option names for almost all Picard tools, without having to go through and manually change everything.  I haven't looked closely at Barclay, but if you don't want to code this into Barclay, one possibility might be to let users of Barclay implement an interface on the tools that takes in a `Field` and it's annotation instance and spit back the long name to use?  Then picard could implement that to use auto-translated field names _unless_ the annotation overrides that with a specific name.

**Comment by magicDGS** [comment]
@droazen, I have a question regarding the long argument convention: I have some toolkits that will use the Picard and/or GATK4 engine, but they have mostly a camelCase convention. If the strict check in the GATK engine will be on by default, is there any possibility that this could be switch off somehow by clients that want to use the engine but requires a different convention? In addition, if the plan is to switch on this convention in Picard, it will be nice to be able to switch on/off this...

Thank you very much!

**Comment by yfarjoun** [comment]
I agree that it would be good to have POSIX-picard use lower-case-dashed arguments that would be auto-translated from the actual variable name.  

We could also do it the other way, by going in and modifying the variable names in picard (manually) and then change the legacy parser to present and use the UPPER_CASED_UNDERSCORED version of the variable so that the long-term will have no auto conversion, and the short-term, legacy will have no break in usage.

**Comment by magicDGS** [comment]
@tfenne and @lbergelsone, you may be interested in having a look to the discussion in https://github.com/broadinstitute/barclay/issues/47, because I did the first draft in one of my Barclay branches to support Markdown in the CLI annotations and in the javadoc to be rendered as HTML in the docgen and to plain-text in the terminal. I'm using [commonmark-java](https://github.com/atlassian/commonmark-java) for it, so maybe it will be helpful also for fgbio.

**Comment by kbergin** [comment]
@cmnbroad Is this work complete?

**Comment by cmnbroad** [comment]
@kbergin I don't think we should close this yet - its blocked by the next phase of this transition, which is https://github.com/broadinstitute/picard/issues/1120.
